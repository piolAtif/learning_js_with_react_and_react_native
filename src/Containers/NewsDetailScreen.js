import React, { Component } from 'react'
import { View, Text, Image, StyleSheet } from 'react-native'

export default class NewsDetailScreen extends Component {
  constructor (props) {
    super(props)
    const news = props.navigation.getParam('news')
    this.state = {
      news
    }
  }

  render () {
    const { news } = this.state
    return (
      <View style={styles.container}>
        <Image style={styles.image} source={{ uri: news.image }} />
        <View style={styles.info}>
          <Text>{news.description}</Text>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  image: {
    height: 200,
    width: '100%',
    alignSelf: 'center'
  },
  info: {
    padding: 10
  }
})
