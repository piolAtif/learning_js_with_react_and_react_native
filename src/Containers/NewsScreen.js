import React, { Component } from 'react'
import { View, FlatList, TouchableOpacity } from 'react-native'
import NewsItem from '../Components/NewsItem'
import CategoryItem from '../Components/CategoryItem'
import Icon from 'react-native-vector-icons/FontAwesome'

const categories = [
  {
    id: 0,
    name: 'All'
  },
  {
    id: 1,
    name: 'Science'
  },
  {
    id: 2,
    name: 'World'
  },
  {
    id: 3,
    name: 'Tech'
  },
  {
    id: 4,
    name: 'Buzz'
  }
]

const news = [
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 2,
    love: 0
  },
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 2,
    love: 0
  },
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 2,
    love: 0
  },
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 1,
    love: 0
  },
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 4,
    love: 0
  },
  {
    image: 'https://cdn.sindonews.net/dyn/620/content/2019/03/28/185/1390809/bunda-ini-cara-membuat-ketoprak-ala-rumahan-Dfx.jpg',
    title: 'Ketoprak Nyangkut',
    description: 'Ketoprak ada yang nyangkut di dinding dan membuat kita klepek-klepek',
    categoryId: 4,
    love: 0
  }
]

export default class NewsScreen extends Component {
  static navigationOptions = ({ navigation }) => {
    return {
      title: 'News Screen',
      headerLeft: (
        <TouchableOpacity testID='leftHeader' onPress={() => navigation.navigate('ProfileScreen')} style={{ padding: 10 }}>
          <Icon size={18} name='user' />
        </TouchableOpacity>
      )
    }
  }

  state = {
    news,
    categories,
    selectedCategory: categories[0]
  }

  changeCategory = (category) => {
    this.setState({
      selectedCategory: category
    })
  }

  renderNews = () => {
    const { selectedCategory, news, categories } = this.state
    if (selectedCategory === categories[0]) {
      return news
    }
    return news.filter((item) => item.categoryId === selectedCategory.id)
  }

  goToNewsDetail = (selectedNews) => {
    const { navigation } = this.props
    navigation.navigate('NewsDetailScreen', { news: selectedNews, loveNews: this.loveNews })
  }

  loveNews = (lovedNews) => {
    const { news } = this.state
    console.log('RWP', lovedNews)
    this.setState({
      news: news.map((newsItem) => lovedNews === newsItem ? { ...newsItem, love: newsItem.love + 1 } : newsItem)
    })
  }

  render () {
    const { selectedCategory, categories } = this.state

    return (
      <View style={{ flex: 1 }}>
        <View style={{ height: 60, justifyContent: 'center' }}>
          <FlatList
            horizontal
            data={categories}
            renderItem={({ item }) =>
              <CategoryItem
                onPress={this.changeCategory}
                category={item}
                active={this.state.selectedCategory === item}
              />
            }
            keyExtractor={(item, index) => index.toString()}
            extraData={selectedCategory}
          />
        </View>
        <FlatList
          testID='newsList'
          data={this.renderNews()}
          renderItem={({ item }) =>
            <NewsItem news={item} onPress={this.goToNewsDetail} />
          }
          keyExtractor={(item, index) => index.toString()}
        />
      </View>
    )
  }
}
