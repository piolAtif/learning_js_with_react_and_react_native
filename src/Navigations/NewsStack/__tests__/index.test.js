import React from 'react'
import { createStackNavigator } from 'react-navigation'
import NewsScreen from '../../../Containers/NewsScreen'
import NewsDetailScreen from '../../../Containers/NewsDetailScreen'

jest.mock('react-navigation', () => ({
  createStackNavigator: jest.fn()
}))

describe('NewsStack', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  describe('', () => {
    it('should ', () => {
      require('../')
      const expectedNavigationStack = {
        NewsScreen: {
          screen: NewsScreen
        },
        NewsDetailScreen: {
          screen: NewsDetailScreen
        }
      }
      expect(createStackNavigator).toHaveBeenCalledWith(expectedNavigationStack)
    })
  })

  describe('ra', () => {
    it('should ', () => {
      require('../')
      const hilih = createStackNavigator.mock.calls[0][0]
      expect(hilih.toJSON()).toMatchSnapshot()
    })
  })
})
