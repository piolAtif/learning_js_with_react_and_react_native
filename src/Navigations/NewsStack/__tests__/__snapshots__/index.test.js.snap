// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsStack ra should  1`] = `
Object {
  "NewsDetailScreen": Object {
    "navigationOptions": [Function],
    "screen": [Function],
  },
  "NewsScreen": Object {
    "navigationOptions": Object {
      "title": "News Screen",
    },
    "screen": [Function],
  },
}
`;

exports[`NewsStack should  1`] = `undefined`;
